### PR TITLE
Midijobfaktor calculation in 2004

### LINKED ISSUE #643: BUG: Midijobfaktor calculation in 2004
### Bug description

When computing social security contributions for 2004 I encountered a bug regarding the calculation of the Midijobfaktor. The function `midijob_faktor_f_mit_minijob_steuerpauschale` is active in 2004 and throws out an error in this line: 

```python
    allg_sozialv_beitr = (
        sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_rentenv"]
->        + sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_pflegev"]["standard"]
        + sozialv_beitr_params["beitr_satz_jahresanfang"]["arbeitsl_v"]
    )
```

This happens because `sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_pflegev"]` is a scalar, the distinction between the standard contribution rate and the extra contribution for people without children exists since 2005.

#### Solution

Implement a type check of `sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_pflegev"]` that returns the correct object.

### PR DESCRIPTION
**Pull request by MImmesberger**
Fixes #643

### CODE REVIEW

**Comment by MImmesberger** [comment]
I implemented 3 type checks to get around the problem. But I'm open to other suggestions how to handle this, I'm not sure whether this solution is very elegant.

File: src/_gettsim/social_insurance_contributions/eink_grenzen.py [hunk 1]
```diff
@@ -158,10 +158,18 @@ def midijob_faktor_f_mit_minijob_steuerpauschale(
     # Therefore sum the contributions which are the same for employee and employer
     allg_sozialv_beitr = (
         sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_rentenv"]
-        + sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_pflegev"]["standard"]
         + sozialv_beitr_params["beitr_satz_jahresanfang"]["arbeitsl_v"]
     )
 
+    if isinstance(sozialv_beitr_params["beitr_satz_jahresanfang"]["ges_pflegev"], dict):
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/eink_grenzen.py:164]: Please get rid of this statement by making two functions out of `midijob_faktor_f_mit_minijob_steuerpauschale` according to the regime.

File: src/_gettsim/social_insurance_contributions/ges_pflegev.py [hunk 2]
```diff
@@ -455,10 +455,15 @@ def _ges_pflegev_beitr_midijob_sum_arbeitn_arbeitg_m(
 
     """
 
-    gesamtbeitrag_midijob_m = midijob_bemessungsentgelt_m * (
-        ges_pflegev_beitr_satz
-        + sozialv_beitr_params["beitr_satz"]["ges_pflegev"]["standard"]
-    )
+    if isinstance(sozialv_beitr_params["beitr_satz"]["ges_pflegev"], dict):
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/ges_pflegev.py:458]: Same as above.

File: src/_gettsim/social_insurance_contributions/ges_pflegev.py [hunk 3]
```diff
@@ -485,7 +490,12 @@ def _ges_pflegev_beitr_midijob_arbeitg_m_anteil_bruttolohn(
     -------
 
     """
-    out = bruttolohn_m * sozialv_beitr_params["beitr_satz"]["ges_pflegev"]["standard"]
+    if isinstance(sozialv_beitr_params["beitr_satz"]["ges_pflegev"], dict):
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/ges_pflegev.py:493]: Once more.

**Review by hmgaudecker — COMMENTED** [review]
Looks great, but we should always try to make the functions as simple as possible and rely more on `@dates_active`

**Comment by MImmesberger** [comment]
Thanks! I implemented the requested changes.
I just want to point out that the functions are being split up because of technical reasons, not because the law changed. But I agree that simplicity is more important.

**Comment by MImmesberger** [comment]
I adjusted quite a few docstrings so I don't want to merge without someone taking a quick look. Can you do that @hmgaudecker?

File: src/_gettsim/social_insurance_contributions/beitr_bemess_grenzen.py [hunk 4]
```diff
@@ -23,7 +23,7 @@ def _ges_rentenv_beitr_bemess_grenze_m(
 def _ges_krankenv_beitr_bemess_grenze_m(
     wohnort_ost: bool, sozialv_beitr_params: dict
 ) -> float:
-    """Calculating the income threshold up to which health insurance payments apply.
+    """Icome threshold up to which health insurance payments apply.
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/beitr_bemess_grenzen.py:26]:
```suggestion
    """Income threshold up to which health insurance payments apply.
```

File: src/_gettsim/social_insurance_contributions/eink_grenzen.py [hunk 5]
```diff
@@ -121,19 +121,82 @@ def in_gleitzone(
...
+    end="2004-12-31",
+    change_name="midijob_faktor_f",
+)
+@add_rounding_spec(params_key="sozialv_beitr")
+def midijob_faktor_f_mit_minijob_steuerpauschale_bis_2004(
+    sozialv_beitr_params: dict,
+    _ges_krankenv_beitr_satz_jahresanfang: float,
+    _ges_krankenv_beitr_satz_arbeitg_jahresanfang: float,
+) -> float:
+    """Compute midijob factor until December 2004.
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/eink_grenzen.py:133]:
```suggestion
    """Midijob Faktor F until December 2004.
```

File: src/_gettsim/social_insurance_contributions/eink_grenzen.py [hunk 6]
```diff
@@ -121,19 +121,82 @@ def in_gleitzone(
...
     change_name="midijob_faktor_f",
 )
 @add_rounding_spec(params_key="sozialv_beitr")
-def midijob_faktor_f_mit_minijob_steuerpauschale(
+def midijob_faktor_f_mit_minijob_steuerpauschale_ab_2005(
     sozialv_beitr_params: dict,
     _ges_krankenv_beitr_satz_jahresanfang: float,
     _ges_krankenv_beitr_satz_arbeitg_jahresanfang: float,
 ) -> float:
     """Faktor F which is needed for the calculation of Bemessungsentgelt
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/eink_grenzen.py:196]:
```suggestion
    """Midijob Faktor F between 2005 and September 2025.
```

(remove remainder explaining the formula, make consistent across all 3 docstrings re faktor_f)

File: src/_gettsim/social_insurance_contributions/ges_krankenv.py [hunk 7]
```diff
@@ -189,7 +189,7 @@ def ges_krankenv_beitr_arbeitg_m_mit_midijob(
 def ges_krankenv_zusatzbeitr_satz_from_sonderbeitr_satz(
     sozialv_beitr_params: dict,
 ) -> float:
-    """Calculate the top-up rate of the health care insurance.
+    """Calculate the top-up rate of the health care insurance until December 2014.
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/ges_krankenv.py:192]:
```suggestion
    """Health insurance top-up rate until December 2014.
```

File: src/_gettsim/social_insurance_contributions/ges_krankenv.py [hunk 8]
```diff
@@ -212,7 +212,7 @@ def ges_krankenv_zusatzbeitr_satz_from_sonderbeitr_satz(
 def ges_krankenv_zusatzbeitr_satz_from_mean_zusatzbeitrag(
     sozialv_beitr_params: dict,
 ) -> float:
-    """Calculate the top-up rate of the health care insurance.
+    """Calculate the top-up rate of the health care insurance since January 2015.
```
Comment (hmgaudecker) [inline src/_gettsim/social_insurance_contributions/ges_krankenv.py:215]:
```suggestion
    """Health insurance top-up rate since January 2015.
```

**Review by hmgaudecker — APPROVED** [review]
Thanks a lot, very nice. Just tiny things.

**Pull request merged**